File: CSharp/Eta/Src/Core/Transport/transport/Pool.cs
/*|-----------------------------------------------------------------------------
 *|            This source code is provided under the Apache 2.0 license
 *|  and is provided AS IS with no warranty or guarantee of fit for purpose.
 *|                See the project's LICENSE.md for details.
 *|           Copyright (C) 2022-2023 LSEG. All rights reserved.     
 *|-----------------------------------------------------------------------------
 */

using System.Diagnostics;

namespace LSEG.Eta.Transports
{
    internal class Pool
    {
        private EtaQueue _queue = new EtaQueue();
        internal object PoolOwner { get; private set; }
        internal bool IsSharedPoolBuffer { get; set; }
        internal bool IsProtocolBuffer { get; set; }

        public Pool(object o, bool isSharedPoolBuffer = false)
        {
            PoolOwner = o;
            IsSharedPoolBuffer = isSharedPoolBuffer;
        }

        public virtual EtaNode Poll()
        {
            EtaNode node = _queue.Poll();
            if (node != null)
                node.InPool = false;
            return node;
        }

        public virtual void Add(EtaNode node)
        {
            node.InPool = true;
            node.Pool = this;
            _queue.Add(node);
        }

        public int Size => _queue.Size;

        /* Add a number of EtaNodes from queue to this pool.
         * 
         * queue is the queue containing the nodes.
         * count is the number of nodes to add from queue to this pool.
         * 
         * Returns the number of nodes added to this pool.
         */
        internal int Add(Pool pool, int numToAdd)
        {
            // adds the nodes from queue to this pool
            int numAdded = 0;
            EtaNode tempNode;
            while ((numAdded < numToAdd) && ((tempNode = pool._queue.Poll()) != null))
            {
                tempNode.InPool = false;
                tempNode.Pool = this;
                tempNode.ReturnToPool();
                numAdded++;
            }
            return numAdded;
        }

        /* Try to get numOfNodes number of nodes from this pool and add it to the pool
         * 
         * Returns number of added nodes
         */
        internal int Poll(Pool pool, int numOfNodes)
        {
            Debug.Assert(pool != this);
            int numAdded = 0;
            EtaNode tempNode = Poll();
            while (tempNode != null)
            {
                tempNode.Pool = pool;
                tempNode.ReturnToPool();
                numAdded++;
                if (numAdded == numOfNodes)
                    break;
                tempNode = Poll();
            }
            return numAdded;
        }

        internal void Clear()
        {
            // This method is used during uninitialize.
            EtaNode node;
            while ((node = Poll()) != null)
            {
                node.Next = null;
            }

            IsSharedPoolBuffer = false;
            IsProtocolBuffer = false;
            PoolOwner = null;
        }
    }
}

File: Java/Eta/Core/src/main/java/com/refinitiv/eta/transport/Pool.java
/*|-----------------------------------------------------------------------------
 *|            This source code is provided under the Apache 2.0 license
 *|  and is provided AS IS with no warranty or guarantee of fit for purpose.
 *|                See the project's LICENSE.md for details.
 *|           Copyright (C) 2020,2025 LSEG. All rights reserved.
 *|-----------------------------------------------------------------------------
 */

package com.refinitiv.eta.transport;

class Pool 
{
    final EtaQueue _queue = new EtaQueue();
    Object _poolOwner;
    boolean _isSharedPoolBuffer;
    boolean _isProtocolBuffer;

    Pool(Object o)
    {
        _poolOwner = o;
    }

    EtaNode poll()
    {
        EtaNode node = _queue.poll();
        if (node != null)
            node._inPool = false;
        return node;
    }

    void add(EtaNode node)
    {
        node._inPool = true;
        _queue.add(node);
    }

    int size()
    {
        return _queue.size();
    }

    /* Add a number of EtaNodes from queue to this pool.
     * 
     * queue is the queue containing the nodes.
     * count is the number of nodes to add from queue to this pool.
     * 
     * Returns the number of nodes added to this pool.
     */
    int add(Pool pool, int numToAdd)
    {
        // adds the nodes from queue to this pool
        int numAdded = 0;
        EtaNode tempNode = null;
        while ((numAdded < numToAdd) && ((tempNode = pool._queue.poll()) != null))
        {
            tempNode._inPool = false;
            tempNode.pool(this);
            tempNode.returnToPool();
            numAdded++;
        }
        return numAdded;
    }

    /* Try to get numOfNodes number of nodes from this pool and add it to the pool
     * 
     * Returns number of added nodes
     */
    int poll(Pool pool, int numOfNodes)
    {
        assert (pool != this);
        int numAdded = 0;
        EtaNode tempNode = poll();
        while (tempNode != null)
        {
            tempNode.pool(pool);
            tempNode.returnToPool();
            numAdded++;
            if (numAdded == numOfNodes)
                break;
            tempNode = poll();
        }
        return numAdded;
    }

    void clear()
    {
        // This method is used during uninitialize.
        EtaNode node = null;
        while ((node = poll()) != null)
        {
            node.next(null);
        }
        _isSharedPoolBuffer = false;
        _isProtocolBuffer = false;
    }

}
